rockcraft/extensions/paas_flask.py

```diff
@@ -0,0 +1,105 @@
+
+"""An experimental extension for the Juju PaaS Flask framework."""
+
+import os
+from typing import Any, Dict, Optional, Tuple
+
+from overrides import override
+
+from .extension import Extension
+
+
+class PaasFlask(Extension):
+    """An extension to enable Flask support within the Juju PaaS ecosystem.
+
+    This extension is a part of the Juju PaaS ecosystem and facilitates the deployment of
+    Flask-based applications.
+    """
+
+    @staticmethod
+    @override
+    def get_supported_bases() -> Tuple[str, ...]:
+        """Return supported bases."""
+        return "bare", "ubuntu:22.04"
+
+    @staticmethod
+    @override
+    def is_experimental(base: Optional[str]) -> bool:
+        """Check if the extension is in an experimental state."""
+        return True
+
+    @override
+    def get_root_snippet(self) -> Dict[str, Any]:
+        """Fill in some default root components for Flask.
+
+        Default values:
+          - run_user: _daemon_
+          - build-base: ubuntu:22.04 (only if user specify bare without a build-base)
+          - platform: amd64
+        """
+        snippet: Dict[str, Any] = {}
+        if "run_user" not in self.yaml_data:
+            snippet["run_user"] = "_daemon_"
+        if (
+            "build-base" not in self.yaml_data
+            and self.yaml_data.get("base", "bare") == "bare"
+        ):
+            snippet["build-base"] = "ubuntu:22.04"
+        if "platforms" not in self.yaml_data:
+            snippet["platforms"] = {"amd64": {}}
+        return snippet
+
+    @override
+    def get_part_snippet(self) -> Dict[str, Any]:
+        """Return the part snippet to apply to existing parts."""
+        return {}
+
+    @override
+    def get_parts_snippet(self) -> Dict[str, Any]:
+        """Create necessary parts to facilitate the flask application.
+
+        Parts added:
+            - flask-extension/dependencies: install Python dependencies
+            - flask-extension/install-app: copy the flask project into the OCI image
+        """
+        python_requirements = []
+        if (self.project_root / "requirements.txt").exists():
+            python_requirements.append("requirements.txt")
+        ignores = [".git", "node_modules"]
+        source_files = [
+            f
+            for f in os.listdir(self.project_root)
+            if f not in ignores and not f.endswith(".rock")
+        ]
+        renaming_map = {f: os.path.join("srv/flask/app", f) for f in source_files}
+        snippet = {
+            "paas-flask/dependencies": {
+                "plugin": "python",
+                "stage-packages": ["python3-venv"],
+                "source": ".",
+                "python-packages": ["gunicorn"],
+                "python-requirements": python_requirements,
+            },
+            "paas-flask/install-app": {
+                "plugin": "dump",
+                "source": ".",
+                "organize": renaming_map,
+                "stage": list(renaming_map.values()),
```

Comment:
how is this supposed to work? Were we not supposed to use the npm target for this?

We should add some minimal documentation on how to work with this

Reply:
This approach aligns with the web team's requirements, as they prefer to construct the static assets themselves before executing the `rockcraft pack` command. And let rockcraft should incorporate the built artifact into the OCI image without making any modifications. I will add some comments about this.

Reply:
For flask, developers don't use npm for static assets since flask is a server side rendered framework primarily. Hence we decided not to add support for `npm` based static assets in the flask extension. For others it could make sense or those assets should be served, e.g., using a content cache instead or dedicated 12-factor extension